File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Helpers/TemplateHelper.kt
package com.abracode.actionui.Helpers

import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.ProvidableCompositionLocal
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Modifier
import com.abracode.actionui.Common.ActionUIElement
import com.abracode.actionui.Common.ActionUIModel
import com.abracode.actionui.Common.ActionUIRegistry
import com.abracode.actionui.Common.LocalActionUILogger
import com.abracode.actionui.Common.LocalWindowModel
import com.abracode.actionui.Common.applyCommonProperties
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * Shared infrastructure for the data-driven repeater (`template`) containers -
 * the Android counterpart of Swift's `Helpers/TemplateHelper.swift`.
 *
 * When a container (`List`, `Section`) declares a [ActionUIElement.template]
 * instead of (or alongside) `children`, it renders one instance of the template
 * per row in `states[`[ActionUIModel.ROWS_STATE_KEY]`]` (a `List<List<String>>`,
 * set through the rows API). String properties in the template carry 1-based
 * column references that are substituted per row:
 *
 *   * `$0`  -> all columns joined with ", "
 *   * `$1`  -> column 0 (first column)
 *   * `$2`  -> column 1
 *   * `$N`  -> column N-1
 *
 * **How the per-row view is built (the load-bearing difference vs. Swift).**
 * Swift sets a `templateContext` on a throw-away `ViewModel` and lets container
 * views re-enter `TemplateHelper` for their children. Android instead does the
 * substitution **eagerly over the whole template subtree** ([substituteElement]
 * walks `properties` + `children` + `content`) and then renders the substituted
 * copy through the normal registry pipeline. A substituted container therefore
 * already holds substituted children, so no per-container special-casing is
 * needed. The one piece of context a leaf still needs - which `List`/`Section`
 * owns it and at which row, for action dispatch - is carried by
 * [LocalTemplateContext] (read by `Button`).
 *
 * **Substitution is single-pass and multi-digit-safe.** A regex replaces every
 * `$N` in one sweep, so a column whose value itself contains `$2` is not
 * re-substituted, and `$12` is read as column 12 (not `$1` then a literal `2`).
 * This is a deliberate refinement over Swift's sequential
 * `replacingOccurrences`; an out-of-range `$N` is left as the literal text, as
 * on Swift.
 */
object TemplateHelper {

    private val COLUMN_REF = Regex("""\$(\d+)""")

    /**
     * Substitutes `$0`/`$1`/`$N` column references in [template] against [row].
     * `$0` joins all columns with ", "; `$N` (1-based) maps to `row[N-1]`; an
     * out-of-range index is left as its literal `$N` text.
     */
    fun substituteString(template: String, row: List<String>): String =
        COLUMN_REF.replace(template) { match ->
            when (val n = match.groupValues[1].toInt()) {
                0 -> row.joinToString(", ")
                else -> row.getOrNull(n - 1) ?: match.value
            }
        }

    /**
     * Recursively rebuilds [element], substituting column references in every
     * **string** primitive (numbers/booleans/null pass through unchanged), so
     * the substitution reaches strings nested in objects and arrays.
     */
    fun substituteJson(element: JsonElement, row: List<String>): JsonElement = when (element) {
        is JsonObject -> JsonObject(element.mapValues { substituteJson(it.value, row) })
        is JsonArray -> JsonArray(element.map { substituteJson(it, row) })
        // JsonPrimitive covers JsonNull too; only string primitives are substituted.
        is JsonPrimitive ->
            if (element.isString) JsonPrimitive(substituteString(element.content, row)) else element
    }

    /**
     * Returns a copy of [template] with column references substituted across its
     * `properties`, `children`, and `content` for the given [row]. The nested
     * `template` field (if any) is left untouched.
     */
    fun substituteElement(template: ActionUIElement, row: List<String>): ActionUIElement =
        template.copy(
            properties = template.properties?.let { substituteJson(it, row) as JsonObject },
            children = template.children?.map { substituteElement(it, row) },
            content = template.content?.let { substituteElement(it, row) },
        )

    /**
     * Renders one template instance for [row]. Substitutes the subtree, looks up
     * the (substituted) root builder, provides a [LocalTemplateContext] so a
     * `Button` in the template dispatches with the owning container's id as
     * `viewID` and [rowIndex] as `viewPartID` (the Swift action convention), and
     * builds it through the registry with the inherited text-style environment
     * and the universal modifiers. Renders nothing if the type is unregistered.
     */
    @Composable
    fun BuildTemplateRow(template: ActionUIElement, row: List<String>, parentID: Int, rowIndex: Int) {
        val logger = LocalActionUILogger.current
        val substituted = substituteElement(template, row)
        val builder = ActionUIRegistry.lookup(substituted.type) ?: return
        CompositionLocalProvider(LocalTemplateContext provides TemplateContext(parentID, rowIndex)) {
            ProvideTextStyleEnvironment(substituted.properties, logger) {
                builder.BuildView(substituted, Modifier.applyCommonProperties(substituted.properties, logger))
            }
        }
    }
}

/**
 * The current template row's context, or `null` outside a template. Carries the
 * owning container's element id and the 0-based row index so a `Button` inside a
 * template dispatches its action with `viewID = parentID`, `viewPartID = rowIndex`
 * - the same convention as Swift's `TemplateContext`.
 */
data class TemplateContext(val parentID: Int, val rowIndex: Int)

/** CompositionLocal carrying the active [TemplateContext]; `null` outside a template row. */
val LocalTemplateContext: ProvidableCompositionLocal<TemplateContext?> =
    staticCompositionLocalOf { null }

/**
 * Reads the rows backing the data-driven element [elementID] from its bound
 * [com.abracode.actionui.Common.ViewModel] in the active window, or an empty
 * list when there is no window, no matching id, or no rows set yet. Reading the
 * snapshot-state map inside composition subscribes the caller, so a host
 * `setElementRows(...)` recomposes the `List` / `Section`.
 */
@Suppress("UNCHECKED_CAST")
@Composable
fun templateRows(elementID: Int): List<List<String>> {
    val viewModel = LocalWindowModel.current?.viewModels?.get(elementID) ?: return emptyList()
    return (viewModel.states[ActionUIModel.ROWS_STATE_KEY] as? List<List<String>>) ?: emptyList()
}

File: ActionUIWeb/src/Helpers/TemplateHelper.js
// TemplateHelper.js — the data-driven repeater (template) infrastructure.
// Web analog of ActionUI/Helpers/TemplateHelper.swift (and the Android
// Helpers/TemplateHelper.kt).
//
// When a container (List now; later Section) declares a `template` subview
// instead of `children`, it renders one instance of the template per row in
// states["content"] ([[String]], set via the rows API). String properties in the
// template carry 1-based column references substituted per row:
//
//   $0  → all columns joined with ", "
//   $1  → column 0 (first column)
//   $2  → column 1
//   $N  → column N-1
//
// How the per-row view is built. Swift carries a templateContext on a throw-away
// ViewModel and lets containers re-enter the helper for their children; the web
// (like Android) instead substitutes the whole subtree eagerly, then builds the
// substituted copy through the normal registry pipeline — so a substituted
// container already holds substituted children, with no per-container special
// case. The one piece a leaf still needs — which container owns it and at which
// row, for action dispatch — rides on a child build context (`ctx.templateContext`),
// read by Button.
//
// Substitution is single-pass and multi-digit-safe (a regex replaces every $N in
// one sweep): a column value that itself contains "$2" is not re-substituted, and
// "$12" reads as column 12, not $1 then a literal 2. An out-of-range $N is left as
// its literal text (matching Swift / Android).
//
// Template instances are not host-addressable: every cloned element's id is
// forced to 0, so the build pipeline seeds/binds nothing (the web's equivalent of
// Swift's throw-away ViewModel) and per-row copies never collide on id. A nested
// `template` subview is left untouched.

import { ActionUIElement } from "../Common/ActionUIElement.js";
import { buildElementView } from "../Common/ActionUIRegistry.js";

const COLUMN_REF = /\$(\d+)/g;

// Substitutes $0 / $1 / $N column references in `str` against `row`.
export function substituteString(str, row) {
    return str.replace(COLUMN_REF, (match, digits) => {
        const n = Number(digits);
        if (n === 0) return row.join(", ");
        return n - 1 < row.length ? row[n - 1] : match;
    });
}

// Substitutes column references in every string, recursing through arrays and
// objects (numbers / booleans / null pass through).
function substituteValue(value, row) {
    if (typeof value === "string") return substituteString(value, row);
    if (Array.isArray(value)) return value.map((entry) => substituteValue(entry, row));
    if (value !== null && typeof value === "object") return substituteProperties(value, row);
    return value;
}

function substituteProperties(properties, row) {
    const out = {};
    for (const [key, value] of Object.entries(properties)) out[key] = substituteValue(value, row);
    return out;
}

// Returns a copy of `element` with column references substituted across its
// properties and routed subviews (children / content / label / rows), and every
// id forced to 0 (template instances aren't host-addressable). A nested
// `template` subview is left untouched.
export function substituteElement(element, row) {
    const properties = substituteProperties(element.properties ?? {}, row);
    let subviews = null;
    if (element.subviews) {
        subviews = {};
        for (const [key, value] of Object.entries(element.subviews)) {
            if (key === "template") {
                subviews[key] = value; // nested repeater prototype: left untouched
            } else if (Array.isArray(value)) {
                // "children" ([ActionUIElement]) or "rows" ([[ActionUIElement]]).
                subviews[key] = value.map((entry) => Array.isArray(entry)
                    ? entry.map((child) => substituteElement(child, row))
                    : substituteElement(entry, row));
            } else {
                subviews[key] = substituteElement(value, row); // "content" / "label"
            }
        }
    }
    return new ActionUIElement(0, element.type, properties, subviews);
}

// Builds one template instance for `row`: substitutes the subtree, then builds it
// through the registry on a child context carrying the template context so a
// Button inside dispatches with the owning container's id as viewID and `rowIndex`
// as viewPartID (the Swift / Android action convention). Returns the HTMLElement.
export function buildTemplateRow(template, row, rowIndex, parentID, ctx) {
    const substituted = substituteElement(template, row);
    const childCtx = { ...ctx, templateContext: { parentID, rowIndex } };
    childCtx.build = (element) => buildElementView(element, childCtx);
    return childCtx.build(substituted);
}

// Wires a container's `template` data-driven mode: one substituted template
// instance per row in states["content"] (set via the rows API), re-rendered in
// place on every rows change. The container `node` keeps its own layout (flex /
// grid); the instances are its direct children. Used by the Lazy stacks/grids
// (List has its own selectable variant in buildDataRows). valueType stays none -
// the rows live in the "content" state, not the value. Returns `node`.
export function renderTemplateRows(node, element, template, ctx) {
    let rows = [];
    const renderRows = () => {
        node.replaceChildren(...rows.map((row, index) =>
            buildTemplateRow(template, row, index, element.id, ctx)));
    };
    if (element.id > 0) {
        ctx.model.bindState(element.id, {
            getState: (key) => (key === "content" ? rows : undefined),
            setState: (key, value) => {
                if (key !== "content") return;
                rows = Array.isArray(value) ? value : [];
                renderRows();
            },
        });
    }
    renderRows();
    return node;
}
